File: TF_Tooling_Web/src/Cursors/TouchlessCursor.ts
import TouchFree from '../TouchFree';
import { TouchFreeInputAction } from '../TouchFreeToolingTypes';

// Class: TouchlessCursor
// This class is a base class for creating custom Touchless cursors for use with TouchFree Tooling.
//
// Override <HandleInputAction> to react to <TouchFreeInputActions> as they are recieved.
//
// For an example of a reactive cursor, see <DotCursor>.
export abstract class TouchlessCursor {
    // Group: Variables

    // Variable: cursor
    // The HTMLElement that represents this cursor
    cursor: HTMLElement | SVGElement | undefined;

    // Variable: enabled
    // Whether the cursor should hide and show depending on hand presence
    enabled: boolean;

    // Variable: cursor
    // Whether the cursor should be visible or not after being enabled
    shouldShow: boolean;

    // Variable: opacityOnHandsLost
    // The opacity of the cursor when hands are lost
    protected opacityOnHandsLost = 1;

    // Group: Functions

    // Function: constructor
    // Registers the Cursor for updates from the <InputActionManager>
    //
    // If you intend to make use of the <WebInputController>, make sure that _cursor has the
    // "touchfree-cursor" class. This prevents it blocking other elements from recieving events.
    constructor(_cursor: HTMLElement | SVGElement | undefined) {
        TouchFree.RegisterEventCallback('TransmitInputAction', this.HandleInputAction.bind(this));

        this.cursor = _cursor;
        this.enabled = true;
        this.shouldShow = true;
    }

    // Function: UpdateCursor
    // Sets the position of the cursor, should be run after <HandleInputAction>.
    protected UpdateCursor(_inputAction: TouchFreeInputAction): void {
        if (this.cursor) {
            let width = this.cursor.clientWidth;
            let height = this.cursor.clientHeight;
            if (this.cursor instanceof HTMLElement) {
                [width, height] = this.GetDimensions(this.cursor);
            }

            this.cursor.style.left = _inputAction.CursorPosition[0] - width / 2 + 'px';
            this.cursor.style.top = _inputAction.CursorPosition[1] - height / 2 + 'px';
        }
    }

    protected GetDimensions(cursor: HTMLElement): [number, number] {
        if (cursor.style.width && cursor.style.height) {
            const getFloat = (dimension: string) => parseFloat(dimension.replace('px', ''));
            return [getFloat(cursor.style.width), getFloat(cursor.style.height)];
        }

        const newCursor = cursor as HTMLImageElement;
        return [newCursor.width, newCursor.height];
    }

    // Function: HandleInputAction
    // The core of the logic for Cursors, this is invoked with each <TouchFreeInputAction> as
    // they are received. Override this function to implement cursor behaviour in response.
    //
    // Parameters:
    //    _inputAction - The latest input action received from TouchFree Service.
    protected HandleInputAction(_inputAction: TouchFreeInputAction): void {
        this.UpdateCursor(_inputAction);
    }

    // Function: ShowCursor
    // Used to make the cursor visible
    ShowCursor(): void {
        this.shouldShow = true;
        if (this.enabled) {
            this.SetCursorOpacity(this.opacityOnHandsLost);
        }
    }

    // Function: HideCursor
    // Used to make the cursor invisible
    HideCursor(): void {
        if (this.shouldShow) {
            // If opacity is NaN or 0 then set it to be 1
            this.opacityOnHandsLost = Number(this.cursor?.style.opacity) || 1;
        }
        this.shouldShow = false;
        this.SetCursorOpacity(0);
    }

    // Function: EnableCursor
    // Used to enable the cursor so that it will show if hands are present
    EnableCursor(): void {
        this.enabled = true;
        if (this.shouldShow) {
            this.opacityOnHandsLost = 1;
            this.ShowCursor();
        }
    }

    // Function: DisableCursor
    // Used to disable the cursor so that it will never show
    DisableCursor(): void {
        this.enabled = false;
        const shouldShowOnEnable = this.shouldShow;
        if (shouldShowOnEnable) {
            this.HideCursor();
        }
        this.shouldShow = shouldShowOnEnable;
    }

    // Function: SetCursorOpacity
    // Used to set the opacity of the cursor
    SetCursorOpacity(opacity: number): void {
        if (this.cursor) {
            this.cursor.style.opacity = opacity.toString();
        }
    }
}

File: TF_Tooling_Unity/Assets/Tooling/Scripts/Cursors/TouchlessCursor.cs
using UnityEngine;
using Ultraleap.TouchFree.Tooling.Connection;

namespace Ultraleap.TouchFree.Tooling.Cursors
{
    // Class: TouchlessCursor
    // This class is a base class for creating custom Touchless cursors for use with
    // TouchFree Tooling.
    //
    // Override <HandleInputAction> to react to <InputActions> as they are recieved.
    //
    // For an example of a reactive cursor, see <DotCursor>.
    public class TouchlessCursor : MonoBehaviour
    {
        // Group: Variables

        // Variable: cursorTransform
        // The transform for the image presented by this cursor
        public RectTransform cursorTransform;
        protected Vector2 targetPos;

        // Variable: cursorSize
        // The standard size of the cursor
        public float cursorSize = 0.25f;

        // Variable: cursorRingThickness
        // The thickness of the cursor ring (if it has one)
        public float cursorRingThickness = 1.5f;

        // Variable: minRingThickness
        // The minimum thickness the ring can be.
        public float minRingThickness = 1.5f;

        // Variable: maxRingThickness
        // The maximum thickness the ring can be.
        public float maxRingThickness = 10;

        public Color primaryColor
        {
            get { return _primaryColor; }
        }
        [SerializeField] protected Color _primaryColor = new Color(1, 1, 1, 1);

        public Color secondaryColor
        {
            get { return _secondaryColor; }
        }
        [SerializeField] protected Color _secondaryColor = new Color(1, 1, 1, 1);

        public Color tertiaryColor
        {
            get { return _tertiaryColor; }
        }
        [SerializeField] protected Color _tertiaryColor = new Color(0, 0, 0, 1);

        // Group: MonoBehaviour Overrides

        // Function: Update
        // Runs on Unity's update loop to keep the attached Cursor at the position
        // last stored in <HandleInputAction>
        protected virtual void Update()
        {
            cursorTransform.anchoredPosition = targetPos;
        }

        // Function: OnEnable
        // Initialises & displays the cursor to its default state when the scene is fully loaded.
        // Also registers the Cursor for updates from the <InputActionManager>
        protected virtual void OnEnable()
        {
            InputActionManager.TransmitInputAction += HandleInputAction;
            InitialiseCursor();
        }

        // Function: OnDisable
        // Deregisters the Cursor so it no longer recieves updates from the
        // <InputActionManager>
        protected virtual void OnDisable()
        {
            InputActionManager.TransmitInputAction -= HandleInputAction;
        }

        // Group: Functions

        // Function: HandleInputAction
        // The core of the logic for Cursors, this is invoked with each <InputAction> as
        // they are recieved. Override this function to implement cursor behaviour in response.
        //
        // Parameters:
        //    _inputData - The latest input action recieved from TouchFree Service.
        protected virtual void HandleInputAction(InputAction _inputData)
        {
            targetPos = _inputData.CursorPosition;
        }

        // Function: InitialiseCursor
        // Invoked when the parent GameObject is enabled.
        // Override this function with any intiialisation steps your cursor needs.
        protected virtual void InitialiseCursor()
        {
        }

        // Function: ShowCursor
        // This ensures that all Cursors will have the ability to be shown or hidden. Be sure to
        // override this function to cover the showing of anything an inheriting cursor uses.
        public virtual void ShowCursor()
        {
            cursorTransform.gameObject.SetActive(true);
        }

        // Function: HideCursor
        // This ensures that all Cursors will have the ability to be shown or hidden. Be sure to
        // override this function to cover the hiding of anything an inheriting cursor uses.
        public virtual void HideCursor()
        {
            cursorTransform.gameObject.SetActive(false);
        }

        // Function: SetColors
        // Used to change the colors of the cursor at runtime. Override it to update the specific
        // UI elements that the colors relate to.
        public virtual void SetColors(Color _primary, Color _secondary, Color _tertiary)
        {
            _primaryColor = _primary;
            _secondaryColor = _secondary;
            _tertiaryColor = _tertiary;
        }

        // Function: SetRingThickness
        // Used to set the <cursorRingThickness> value. Can be overridden to remap or clamp this value.
        // _thickness should range from 0-1 0 being the thinnest and 1 being the thickest.
        public virtual void SetRingThickness(float _thickness)
        {
            cursorRingThickness = Utilities.MapRangeToRange(_thickness, 0, 1, minRingThickness, maxRingThickness);
        }
    }
}